lib/Dialect/LLHD/Transforms/LowerProcesses.cpp

```diff
@@ -0,0 +1,279 @@
+    // aggregate form from which any changes remain apparent.
+    if (isa<hw::ArrayCreateOp, hw::StructCreateOp, comb::ConcatOp>(op))
+      for (auto operand : op->getOperands())
+        markObserved(operand);
+  }
+}
+
+/// Ensure that any value defined outside the process that is used inside the
+/// process is derived entirely from an observed value.
+bool Lowering::allOperandsObserved() {
+  // Collect all ancestor regions such that we can easily check if a value is
+  // defined outside the process.
+  SmallPtrSet<Region *, 4> properAncestors;
+  for (auto *region = processOp->getParentRegion(); region;
+       region = region->getParentRegion())
+    properAncestors.insert(region);
+
+  // Walk all operations under the process and check each operand.
+  auto walkResult = processOp.walk([&](Operation *op) {
+    for (auto operand : op->getOperands()) {
+      // We only care about values defined outside the process.
+      if (!properAncestors.count(operand.getParentRegion()))
+        continue;
+
+      // If the value is observed, all is well.
+      if (isObserved(operand))
+        continue;
+
+      // Otherwise complain and abort.
+      LLVM_DEBUG({
+        llvm::dbgs() << "Skipping process " << processOp.getLoc()
+                     << ": unobserved value ";
+        operand.print(llvm::dbgs(), OpPrintingFlags().skipRegions());
+        llvm::dbgs() << "\n";
+      });
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  return !walkResult.wasInterrupted();
+}
+
+/// Check if a value is observed by the wait op, or all its operands are only
+/// derived from observed values.
+bool Lowering::isObserved(Value value) {
+  // Check if the value is trivially observed.
+  if (observedValues.contains(value))
+    return true;
+
+  // Otherwise get the operation that defines it such that we can check if the
+  // value is derived from purely observed values. If it isn't define by an op,
+  // the value is unobserved.
+  auto *defOp = value.getDefiningOp();
+  if (!defOp)
+    return false;
+
+  // Otherwise visit all ops in the fan-in cone and ensure that they are
+  // observed. If anything value is unobserved, immediately return false.
+  SmallDenseSet<Operation *> seenOps;
+  SmallVector<Operation *> worklist;
+  seenOps.insert(defOp);
+  worklist.push_back(defOp);
+  while (!worklist.empty()) {
+    auto *op = worklist.pop_back_val();
+
+    // Give up on ops with nested regions.
+    if (op->getNumRegions() != 0)
+      return false;
+
+    // Otherwise check that all operands are observed. If we haven't seen an
+    // operand before, and it is not a signal, add it to the worklist to be
+    // checked.
+    for (auto operand : op->getOperands()) {
+      if (observedValues.contains(operand))
+        continue;
+      if (isa<hw::InOutType>(operand.getType()))
+        return false;
+      auto *defOp = operand.getDefiningOp();
```

Comment:
Do we conservatively need to check for no side-effects? Registers should be fine, but not sure about memories.

Reply:
Yeah good point, we probably should stop on side-effecting ops. Let me add that.